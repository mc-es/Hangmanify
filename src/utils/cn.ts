import type { ClassValue } from 'clsx';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * ### Combines conditional Tailwind class names and resolves conflicts
 *
 * @param {...ClassValue[]} inputs - Class names, booleans, objects, or arrays
 * @returns {string} Merged and deduplicated class string
 * @example
 * cn('p-4', isDark && 'bg-black') // 'p-4 bg-black' if isDark is true, otherwise 'p-4'
 */
export const cn = (...inputs: ClassValue[]): string => twMerge(clsx(...inputs));
