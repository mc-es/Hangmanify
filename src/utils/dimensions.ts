import { Dimensions as NativeDimensions, PixelRatio } from 'react-native';

const { width: deviceWidth, height: deviceHeight } = NativeDimensions.get('window');

const guidelineBaseWidth = 375;
const guidelineBaseHeight = 812;
const BLEND_ZONE = 5;

const BREAKPOINTS = {
  '2xs': { size: 320, scale: 0.65 },
  'xs-2xs': { size: 340, scale: 0.7 },
  xs: { size: 360, scale: 0.75 },
  'xs-sm': { size: 375, scale: 0.75 },
  sm: { size: 390, scale: 0.85 },
  'sm-md': { size: 420, scale: 0.9 },
  md: { size: 480, scale: 1 },
  'md-lg': { size: 540, scale: 1.05 },
  lg: { size: 600, scale: 1.1 },
  'lg-xl': { size: 680, scale: 1.15 },
  xl: { size: 760, scale: 1.2 },
  'xl-2xl': { size: 800, scale: 1.3 },
  '2xl': { size: 840, scale: 1.4 },
} as const;

const easeIn = (t: number): number => t * t;
const easeOut = (t: number): number => t * (2 - t);

const getScale = (width: number): number => {
  const breakpoints = Object.values(BREAKPOINTS).sort((x, y) => x.size - y.size);

  if (width <= breakpoints[0].size) return breakpoints[0].scale;
  if (width >= breakpoints[breakpoints.length - 1].size)
    return breakpoints[breakpoints.length - 1].scale;

  for (let i = 0; i < breakpoints.length - 1; i++) {
    const current = breakpoints[i];
    const next = breakpoints[i + 1];

    if (width >= current.size && width < next.size) {
      const scaleRange = next.scale - current.scale;
      const widthRange = next.size - current.size;
      const delta = width - current.size;

      let finalProgress;

      if (delta < BLEND_ZONE) {
        const t = delta / BLEND_ZONE;
        finalProgress = easeIn(t) * (BLEND_ZONE / widthRange);
      } else if (delta > widthRange - BLEND_ZONE) {
        const t = (delta - (widthRange - BLEND_ZONE)) / BLEND_ZONE;
        finalProgress =
          (widthRange - BLEND_ZONE) / widthRange + easeOut(t) * (BLEND_ZONE / widthRange);
      } else finalProgress = delta / widthRange;

      return current.scale + scaleRange * finalProgress;
    }
  }

  return 1;
};

const scale = getScale(deviceWidth);

// horizontal & vertical scale
const hs = (size: number): number => (deviceWidth / guidelineBaseWidth) * size * scale;
const vs = (size: number): number => (deviceHeight / guidelineBaseHeight) * size;

// width & height percent
const wp = (percent: number): number => (deviceWidth * percent) / 100;
const hp = (percent: number): number => (deviceHeight * percent) / 100;

// moderate scale
const ms = (size: number, factor = 0.5): number => size + (hs(size) - size) * factor;

// font scale
const fs = (size: number): number => {
  const fontScale = PixelRatio.getFontScale();
  return Math.round(PixelRatio.roundToNearestPixel(hs(size) * fontScale));
};

export const Dimensions = { hs, vs, wp, hp, ms, fs };
