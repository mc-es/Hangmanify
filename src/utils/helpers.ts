import type { TranslateOptions } from 'i18n-js';

function getNestedValue<T>(obj: T, key: string): unknown {
  return key.split('.').reduce((acc: unknown, curr: string): unknown => {
    if (acc !== null && typeof acc === 'object' && curr in acc)
      return (acc as Record<string, unknown>)[curr];

    return null;
  }, obj as unknown);
}
function interpolate(template: string, options: TranslateOptions): string {
  return template.replace(/{{\s*(\w+)\s*}}/g, (_match, key) =>
    key in options ? String(options[key]) : ''
  );
}

export const helpers = { getNestedValue, interpolate };
