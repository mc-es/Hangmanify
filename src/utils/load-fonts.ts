import * as Font from 'expo-font';

import { APP_FONTS } from 'assets/fonts';

import { FONTS } from 'src/constants/styles';

export const LoadFonts = async (): Promise<void> => {
  try {
    await Font.loadAsync({
      [FONTS.Nunito.Bold]: APP_FONTS.nunito.bold,
      [FONTS.Nunito.Light]: APP_FONTS.nunito.light,
      [FONTS.Nunito.Medium]: APP_FONTS.nunito.medium,
      [FONTS.Nunito.Regular]: APP_FONTS.nunito.regular,
      [FONTS.Nunito.SemiBold]: APP_FONTS.nunito.semibold,
      [FONTS.Poppins.Bold]: APP_FONTS.poppins.bold,
      [FONTS.Poppins.Light]: APP_FONTS.poppins.light,
      [FONTS.Poppins.Medium]: APP_FONTS.poppins.medium,
      [FONTS.Poppins.Regular]: APP_FONTS.poppins.regular,
      [FONTS.Poppins.SemiBold]: APP_FONTS.poppins.semibold,
    });
  } catch (error) {
    console.error('Error loading fonts', error);
  }
};
