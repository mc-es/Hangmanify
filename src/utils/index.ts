import { cn } from './cn';
import { dimensions } from './dimensions';
import { helpers } from './helpers';

export { cn, dimensions, helpers };
