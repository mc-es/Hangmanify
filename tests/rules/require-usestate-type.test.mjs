/**
 * @fileoverview Tests for require-usestate-type rule.
 * @see https://github.com/mc-es/Hangmanify/blob/master/guides/rules/require-usestate-type.md
 */
import rule from 'plugins/rules/require-usestate-type.mjs';

import { TEST_FILENAME_PATH } from './Test';
import { runTest } from './utils.js';

const validCases = [
  {
    name: 'allows string type explicitly set',
    filename: TEST_FILENAME_PATH,
    code: `const [val, setVal] = useState<string>('');`,
  },
  {
    name: 'allows boolean type with React namespace',
    filename: TEST_FILENAME_PATH,
    code: `const [flag, setFlag] = React.useState<boolean>(true);`,
  },
  {
    name: 'allows number type explicitly set',
    filename: TEST_FILENAME_PATH,
    code: `const [count, setCount] = useState<number>(0);`,
  },
  {
    name: 'allows array type explicitly set',
    filename: TEST_FILENAME_PATH,
    code: `const [items, setItems] = useState<string[]>([]);`,
  },
  {
    name: 'allows object type via type alias',
    filename: TEST_FILENAME_PATH,
    code: `
      type User = { id: number; name: string };
      const [user, setUser] = useState<User>({ id: 1, name: 'Alice' });
    `,
  },
  {
    name: 'allows union literal via type alias',
    filename: TEST_FILENAME_PATH,
    code: `
      type Status = 'done' | 'loading';
      const [status, setStatus] = useState<Status>('loading');
    `,
  },
];

const invalidCases = [
  {
    name: 'disallows missing type for string value',
    filename: TEST_FILENAME_PATH,
    code: `const [val, setVal] = useState('');`,
    expectedErrors: 1,
  },
  {
    name: 'disallows missing type with React.useState',
    filename: TEST_FILENAME_PATH,
    code: `const [flag, setFlag] = React.useState(true);`,
    expectedErrors: 1,
  },
  {
    name: 'disallows missing type for number value',
    filename: TEST_FILENAME_PATH,
    code: `const [count, setCount] = useState(0);`,
    expectedErrors: 1,
  },
  {
    name: 'disallows missing type for array value',
    filename: TEST_FILENAME_PATH,
    code: `const [items, setItems] = useState([]);`,
    expectedErrors: 1,
  },
  {
    name: 'disallows missing type for object literal',
    filename: TEST_FILENAME_PATH,
    code: `const [user, setUser] = useState({ id: 1, name: 'Alice' });`,
    expectedErrors: 1,
  },
  {
    name: 'disallows missing type for union literal value',
    filename: TEST_FILENAME_PATH,
    code: `const [status, setStatus] = useState('loading');`,
    expectedErrors: 1,
  },
];

runTest('require-usestate-type', rule, validCases, invalidCases);
