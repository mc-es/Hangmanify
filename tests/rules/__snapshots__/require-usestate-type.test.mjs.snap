// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`require-usestate-type rule -> invalid cases -> disallows missing type for array value 1`] = `
[
  {
    "column": 27,
    "endColumn": 39,
    "endLine": 1,
    "line": 1,
    "message": "'useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;

exports[`require-usestate-type rule -> invalid cases -> disallows missing type for number value 1`] = `
[
  {
    "column": 27,
    "endColumn": 38,
    "endLine": 1,
    "line": 1,
    "message": "'useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;

exports[`require-usestate-type rule -> invalid cases -> disallows missing type for object literal 1`] = `
[
  {
    "column": 25,
    "endColumn": 59,
    "endLine": 1,
    "line": 1,
    "message": "'useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;

exports[`require-usestate-type rule -> invalid cases -> disallows missing type for string value 1`] = `
[
  {
    "column": 23,
    "endColumn": 35,
    "endLine": 1,
    "line": 1,
    "message": "'useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;

exports[`require-usestate-type rule -> invalid cases -> disallows missing type for union literal value 1`] = `
[
  {
    "column": 29,
    "endColumn": 48,
    "endLine": 1,
    "line": 1,
    "message": "'useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;

exports[`require-usestate-type rule -> invalid cases -> disallows missing type with React.useState 1`] = `
[
  {
    "column": 25,
    "endColumn": 45,
    "endLine": 1,
    "line": 1,
    "message": "'React.useState' should always have a type argument.",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/require-usestate-type",
    "severity": 2,
  },
]
`;
